### PR TITLE
Modify CI to track ros testing repo

### PR DESCRIPTION
**Pull request by CihatAltiparmak**
To prevent the errors stated in this PR ( https://github.com/ros2/rmw_zenoh/pull/255 ) , i have refactored the build CI so that it can be tracked the modifications, API changes etc. for rolling, jazzy and iron continuously. I don't know what type of approach this team has, however i wanted to come up with some option. It's ready to your service. It doesn't matter if you accepted or not. Btw to just modify container except this PR's idea can be another option. I can do it too.

### CODE REVIEW

**Comment by MichaelOrlov** [comment]
@Yadunund @clalancette Could you please review this PR?

**Review by Yadunund — CHANGES_REQUESTED** [review]
Thanks for kicking off this initiative! 

I feel that the using binaries from the `testing` repo may not highlight issues soon enough. It relies on someone blooming changes from the core repos and then waiting for  the rosdistro PR to be merged to kick-off the build jobs.

On the other hand we expect core packages for released distros like jazzy, iron etc to be fairly stable so we don't need to test with the latest source code or binaries from the test repo.

What do you think of the following:
- Update the `build.yaml` workflow to rely on [setup-ros](https://github.com/ros-tooling/setup-ros).
  - If we're testing for `rolling`, the job will build `--packages-up-to rmw_zenoh_cpp` from **source**. 
  - Else, we configure `setup-ros` to use regular ROS 2 core binaries and build upto `rmw_zenoh_cpp`
- Configure the build job to run on a nightly schedule using a [cron config](https://github.com/open-rmf/rmf/blob/100453f07c74b48babc7333a25c6db342dd187b1/.github/workflows/build.yaml#L5-L6). This way the maintainers will be notified if any breaking changes get merged in core upstream repos.

**Comment by clalancette** [comment]
Agreed with what @Yadunund laid out above.

**Comment by ahcorde** [comment]
I made the changes @Yadunund 

 - source build for rolling `build / build_and_test_source_rolling `
 - binary build for jazzy `build / build_and_test_binaries_jazzy `
 - binary build for iron `build / build_and_test_binaries_iron`
 - Added cron too

**Comment by ahcorde** [comment]
[rolling is failing](https://github.com/ros2/rmw_zenoh/actions/runs/10506648470/job/29106714188?pr=256) because of the `localhost_only` issue https://github.com/ros2/rmw_zenoh/pull/255

File: .github/workflows/build.yaml
Comment (Yadunund) [inline .github/workflows/build.yaml]: Instead of having three separate jobs like this, can we retain the `distro` matrix from earlier and setup one job only to run for each distro in the matrix? Within this job, we can have a special `if` condition to check if `distro == rolling` and if so, build from source. If you need access to both the ROS distro version and ubuntu version, the matrix can be defined [like so](https://github.com/open-rmf/rmf_ci_templates/blob/1ae7b4f64644ab2f8724e9a1e1b3d14803caa61e/.github/workflows/reusable_build.yaml#L13-L25).

Also it appears that your changes now pull a `rostooling/setup-ros-docker:ubuntu-x` image. How often is this image updated? And why not use `ros-tooling/setup-ros@v0.7` to setup from latest `ubuntu` base image [like so](https://github.com/ros-tooling/setup-ros?tab=readme-ov-file#setting-up-the-worker-installing-system-dependencies-and-ros-linux).

**Review by Yadunund — CHANGES_REQUESTED** [review]
Thanks @ahcorde for the changes. Left a comment to reuse the same job across distros/os.

File: .github/workflows/build.yaml
Comment (Yadunund) [reply .github/workflows/build.yaml]: Here is an example of using an `if` condition within the action: https://github.com/open-rmf/rmf_ci_templates/blob/1ae7b4f64644ab2f8724e9a1e1b3d14803caa61e/.github/workflows/reusable_build.yaml#L53

File: .github/workflows/build.yaml
Comment (Yadunund) [inline .github/workflows/build.yaml]: We can't alwayse use `noble` since iron is distributed on `jammy`. Please see my previous comment on how the matrix can be updated to include distro and os. https://github.com/ros2/rmw_zenoh/pull/256#discussion_r1727744706

Not addressed form the same comment:
* Using If condition to include `vcs-repo-file-url` 
* Using ros-tooling/setup-ros@v0.7 instead of the `rostooling/setup-ros-docker:ubuntu` image

**Review by Yadunund — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yaml
Comment (CihatAltiparmak) [reply .github/workflows/build.yaml]: Firstly sorry for some commits, but would using separate jobs for source and binaries be good instead of using if?  I am yet to complete my work. I will ping you.

Comment (Yadunund) [reply .github/workflows/build.yaml]: I'd prefer to have everything in the same job. The CI here is a temporary measure until we start running the usual CI jobs using the ROS buildfarm.

Comment (CihatAltiparmak) [reply .github/workflows/build.yaml]: I've applied, @Yadunund and i've readded the fail-fast param which @ahcorde has overlooked by deleting this. Would you review in your available time? (Inspired by https://github.com/ros-tooling/setup-ros/blob/a6ce30ecca1e5dcc10ae5e6a44fe2169115bf852/README.md?plain=1#L249-L252 and https://github.com/ros2/ros2_tracing/blob/a866b9c701311bc8200a00c949b4a0fff9803777/.github/workflows/test.yml#L68) Btw if you want, we can merge `rolling` branch to this branch so that all CI can pass.

Comment (ahcorde) [reply .github/workflows/build.yaml]: thank you for the changes @CihatAltiparmak

**Review by ahcorde — APPROVED** [review]
I will wait for a final review from @Yadunund

File: .github/workflows/build.yaml
Comment (Yadunund) [inline .github/workflows/build.yaml]: I think we should remove this `DOCKER_IMAGE` flag as it's one extra item to configure.
See commend below for relevant suggestion.

File: .github/workflows/build.yaml
Comment (Yadunund) [inline .github/workflows/build.yaml]: Our CI times have increased  by 2mins by installed ROS binaries for cases where `BUILD_TYPE` is `binary`. Can we use an if condition here to use `ros:${{ matrix. ROS_DISTRO }}-ros-base` here if `matrix.BUILD_TYPE == 'binary'`; else `ubuntu:noble`. We won't require the DOCKER_IMAGE variable anymore.

**Review by Yadunund — CHANGES_REQUESTED** [review]
It looks like CI time for binary jobs has increased by little over 2 mins due to the fact that we're installing ROS binaries as part of the workflow. Left a suggestion for how we could tackle this.

File: .github/workflows/build.yaml
Comment (CihatAltiparmak) [reply .github/workflows/build.yaml]: I've modified based on your suggestion. But i've also had to install coverage tools in case of using binary because of crashing CI. I hope i've decreased CI check time. Finally, my approach is okay with binary builds, but It can create some problems in case that you want to build from iron source because iron fails in `ubuntu-noble` 
While modifying, i've utilized https://github.com/open-rmf/rmf_ci_templates/blob/1ae7b4f64644ab2f8724e9a1e1b3d14803caa61e/.github/workflows/reusable_build.yaml#L56-L59

**Review by Yadunund — APPROVED** [review]
LGTM! Thanks for iterating @CihatAltiparmak and for addressing the build error on Rolling!

**Pull request merged**
